tools/please_go_install/main.go

```diff
@@ -108,114 +104,70 @@ func (g *pkgGraph) compile(from []string, target string) {
-		}
-	}
-	binary := targetPackage.Name == "main"
-	if anyCGO {
-		goToolCGOCompile(target, binary, pkgDir, srcs, cgoSrcs)
+	if len(pkg.CgoFiles) > 0 {
+		goToolCGOCompile(target, pkg)
 	} else {
-		goToolCompile(target, binary, srcs) // output the package as ready to be compiled
+		goToolCompile(target, pkg) // output the package as ready to be compiled
 	}
 	g.pkgs[target] = true
 }
 
-func goToolCompile(target string, binary bool, srcs []string) {
+func goToolCompile(target string, pkg *build.Package) {
+	fullSrcPaths := make([]string, len(pkg.GoFiles))
+	for i, src := range pkg.GoFiles {
+		fullSrcPaths[i] = filepath.Join(pkg.Dir, src)
+	}
 	out := fmt.Sprintf("%s/%s.a", opts.Out, target)
 
 	prepOutDir := fmt.Sprintf("mkdir -p %s", filepath.Dir(out))
-	compile := fmt.Sprintf("%s tool compile -pack -complete -importcfg %s -o %s %s", opts.GoTool, opts.ImportConfig, out, strings.Join(srcs, " "))
+	compile := fmt.Sprintf("%s tool compile -pack -complete -importcfg %s -o %s %s", opts.GoTool, opts.ImportConfig, out, strings.Join(fullSrcPaths, " "))
 	updateImportCfg := fmt.Sprintf("echo \"packagefile %s=%s\" >> %s", target, out, opts.ImportConfig)
 
 	fmt.Println(prepOutDir)
 	fmt.Println(compile)
 	fmt.Println(updateImportCfg)
 
-	if binary {
+	if pkg.IsCommand() {
 		goToolLink(out)
 	}
 }
 
-func goToolCGOCompile(target string, binary bool, pkgDir string, srcs []string, cgoSrcs []string) {
+func goToolCGOCompile(target string, pkg *build.Package) {
 	out := fmt.Sprintf("%s/%s.a", opts.Out, target)
 
-	// We need to operate out of the package working directory for the cpp compiler to play ball so trim the package dir
-	// from the source paths.
-	for i := range srcs {
-		srcs[i] = strings.TrimPrefix(strings.TrimPrefix(srcs[i], pkgDir), "/")
-	}
-
-	for i := range cgoSrcs {
-		cgoSrcs[i] = strings.TrimPrefix(strings.TrimPrefix(cgoSrcs[i], pkgDir), "/")
-	}
-
 	prepOutDir := fmt.Sprintf("mkdir -p %s", filepath.Dir(out))
-	cdPkgDir := fmt.Sprintf("cd %s", pkgDir)
-	generateCGO := fmt.Sprintf("%s tool cgo %s", opts.GoTool, strings.Join(cgoSrcs, " "))
+	cdPkgDir := fmt.Sprintf("cd %s", pkg.Dir)
+	generateCGO := fmt.Sprintf("%s tool cgo %s", opts.GoTool, strings.Join(pkg.CgoFiles, " "))
 	compileGo := fmt.Sprintf("%s tool compile -pack -importcfg $OLDPWD/%s -o out.a _obj/*.go", opts.GoTool, opts.ImportConfig)
+	// TODO(jpoole): We can use pkg to determine the correct cgo flags to pass to the compiler here
 	compileCGO := fmt.Sprintf("%s -Wno-error -ldl -Wno-unused-parameter -c -I _obj -I . _obj/_cgo_export.c _obj/*.cgo2.c", opts.CCTool)
 	compileC := fmt.Sprintf("%s -Wno-error -ldl -Wno-unused-parameter -c -I _obj -I . *.c", opts.CCTool)
 	mergeArchive := fmt.Sprintf("%s tool pack r out.a *.o ", opts.GoTool)
-	moveArchive := fmt.Sprintf("cd $OLDPWD && mv %s/out.a %s", pkgDir, out)
+	moveArchive := fmt.Sprintf("cd $OLDPWD && mv %s/out.a %s", pkg.Dir, out) //TODO(jpoole): can we just output this in the right place to begin with?
```

Comment:
Good question. AFAIK the go build tool changes directory into the obj dir and runs all build tasks there.

Reply:
Yep pretty sure it does.

Reply:
Defo some room for improvement but i don't think this code is long for the world. We should be using your toolchain approach here.